src/utilities/hooks.ts

```diff
@@ -1,146 +1,161 @@
 import { NavigationAbortError, RouterPushError, RouterRejectionError } from '@/errors'
 import { RouteHookStore } from '@/models/RouteHookStore'
+import { RouterRoutes } from '@/types'
 import { AfterRouteHook, AfterRouteHookResponse, BeforeRouteHook, BeforeRouteHookResponse, RouteHookAbort, RouteHookLifecycle } from '@/types/hooks'
 import { ResolvedRoute } from '@/types/resolved'
 import { RouterReject } from '@/types/router'
-import { RouterPushImplementation } from '@/types/routerPush'
-import { RouterReplaceImplementation } from '@/types/routerReplace'
+import { RegisteredRouterPush, RouterPush } from '@/types/routerPush'
+import { RegisteredRouterReplace, RouterReplace } from '@/types/routerReplace'
 import { getAfterRouteHooksFromRoutes, getBeforeRouteHooksFromRoutes } from '@/utilities/getRouteHooks'
 
+type RouteHookRunners<T extends RouterRoutes> = {
+  runBeforeRouteHooks: RouteHookBeforeRunner<T>,
+  runAfterRouteHooks: RouteHookAfterRunner<T>,
+}
+
 type BeforeContext = {
   to: ResolvedRoute,
   from: ResolvedRoute,
   hooks: RouteHookStore,
 }
 
-export async function runBeforeRouteHooks({ to, from, hooks }: BeforeContext): Promise<BeforeRouteHookResponse> {
-  const { global, component } = hooks
-  const route = getBeforeRouteHooksFromRoutes(to, from)
-
-  const allHooks: BeforeRouteHook[] = [
-    ...global.onBeforeRouteEnter,
-    ...route.onBeforeRouteEnter,
-    ...global.onBeforeRouteUpdate,
-    ...route.onBeforeRouteUpdate,
-    ...component.onBeforeRouteUpdate,
-    ...global.onBeforeRouteLeave,
-    ...route.onBeforeRouteLeave,
-    ...component.onBeforeRouteLeave,
-  ]
-
-  try {
-    const results = allHooks.map(callback => callback(to, {
-      from,
-      reject,
-      push,
-      replace,
-      abort,
-    }))
-
-    await Promise.all(results)
-
-  } catch (error) {
-    if (error instanceof RouterPushError) {
-      return {
-        status: 'PUSH',
-        to: error.to,
-      }
-    }
+type RouteHookBeforeRunner<T extends RouterRoutes> = (context: BeforeContext) => Promise<BeforeRouteHookResponse<T>>
```

Comment:
I didn't like that it requires T to be passed because it cannot be inferred as it is currently written.